### PR TITLE
helm: Update pod affinity for cilium-envoy

### LINKED ISSUE #31149: Should cilium-envoy tolerations differ from cilium-agent?
I noticed in a recent [CI run](https://github.com/cilium/cilium/actions/runs/8146839676/job/22266268950#step:16:442) that there were three instances of the cilium-envoy Pod and only two of the main `cilium` Pod:

```
	 kube-system         cilium-envoy-4kmmg                           1/1     Running    0          4m27s   172.18.0.3   kind-worker2         <none>           <none>
	 kube-system         cilium-envoy-kf7xl                           1/1     Running    0          4m27s   172.18.0.4   kind-worker          <none>           <none>
	 kube-system         cilium-envoy-zjphl                           1/1     Running    0          4m27s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         cilium-node-init-42whq                       1/1     Running    0          4m27s   172.18.0.3   kind-worker2         <none>           <none>
	 kube-system         cilium-node-init-q5m5s                       1/1     Running    0          4m27s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         cilium-node-init-xgskv                       1/1     Running    0          4m27s   172.18.0.4   kind-worker          <none>           <none>
	 kube-system         cilium-operator-f6bc44d98-l75dk              1/1     Running    0          4m27s   172.18.0.3   kind-worker2         <none>           <none>
	 kube-system         cilium-operator-f6bc44d98-sjmv2              1/1     Running    0          4m27s   172.18.0.4   kind-worker          <none>           <none>
	 kube-system         cilium-qv62g                                 0/1     Init:5/7   0          4m27s   172.18.0.4   kind-worker          <none>           <none>
	 kube-system         cilium-xjpqz                                 1/1     Running    0          4m27s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         coredns-76f75df574-kfxd7                     1/1     Running    0          7m35s   10.0.1.121   kind-control-plane   <none>           <none>
	 kube-system         coredns-76f75df574-pjhfd                     1/1     Running    0          7m35s   10.0.1.242   kind-control-plane   <none>           <none>
	 kube-system         etcd-kind-control-plane                      1/1     Running    0          9m18s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         kube-apiserver-kind-control-plane            1/1     Running    0          9m18s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         kube-controller-manager-kind-control-plane   1/1     Running    0          9m18s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         kube-scheduler-kind-control-plane            1/1     Running    0          9m18s   172.18.0.2   kind-control-plane   <none>           <none>
	 kube-system         log-gatherer-bmp7p                           1/1     Running    0          8m48s   172.18.0.3   kind-worker2         <none>           <none>
	 kube-system         log-gatherer-jgmnp                           1/1     Running    0          8m48s   172.18.0.4   kind-worker          <none>           <none>
	 kube-system         log-gatherer-zzb4b                           1/1     Running    0          8m48s   172.18.0.2   kind-control-plane   <none>           <none>
```

Seems like the tolerations are configured separately, but should cilium-envoy run on nodes that don't have cilium running?

### PR DESCRIPTION
**Pull request by sayboras**
This commit is to avoid cilium-envoy running on the node without cilium
agent.

Relates: https://github.com/cilium/cilium/pull/25081, https://github.com/cilium/cilium/pull/30034
Fixes: https://github.com/cilium/cilium/issues/31149
Signed-off-by: Tam Mach <tam.mach@cilium.io>

### CODE REVIEW

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 1]
```diff
@@ -1925,6 +1925,15 @@ envoy:
           labelSelector:
             matchLabels:
               k8s-app: cilium-envoy
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
```
Comment (aanm) [inline install/kubernetes/cilium/values.yaml.tmpl]: I think this should be `requiredDuringSchedulingIgnoredDuringExecution` as we have an example on https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ (grep for the example above the sentence `Creating the two preceding Deployments results in the following cluster layout, where each web server is`)

**Review by aanm — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 2]
```diff
@@ -1925,6 +1925,15 @@ envoy:
           labelSelector:
             matchLabels:
               k8s-app: cilium-envoy
+    podAffinity:
```
Comment (squeed) [inline install/kubernetes/cilium/values.yaml.tmpl:1939]: Why is this in the values and not the daemonset? I would rather have this be in the chart, and then overridable if provided.

**Comment by sayboras** [comment]
Convert to draft to address review comments.

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (sayboras) [reply install/kubernetes/cilium/values.yaml.tmpl:1939]: good point.

Updated to actually enforce cilium-envoy to always have pod-affinity with cilium-agent, as cilium-envoy pods will be just no use without agent.

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 1 above)
Comment (sayboras) [reply install/kubernetes/cilium/values.yaml.tmpl]: Good point :+1:

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (sayboras) [reply install/kubernetes/cilium/values.yaml.tmpl:1939]: Thinking about this again, it's hard to compose different attributes under affinity (e.g. affinity.{podAffinity,podAntiAffinity,nodeAffinity}.{requiredDuringSchedulingIgnoredDuringExecution,preferredDuringSchedulingIgnoredDuringExecution} in a clean way. Hence, I think we can just keep default value in values.yaml as what it is now. Another benefit is consistency across multiple components.

**Review by joamaki — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Pull request merged**